### PR TITLE
RN-748 viz validation

### PR DESCRIPTION
**Pull request by IgorNadj**
### Changes:
- Add viz type selector
- Define initial DashboardItem.config typing (and auto-generated schema)
- Add schema to viz json editor to get validation and autocompletion

---

### Manual release steps:
- [ ] Update docs

### Screenshots:
https://linear.app/bes/issue/RN-748#comment-7d7a49ab

### CODE REVIEW

File: packages/admin-panel/src/VizBuilderApp/App.js [hunk 1]
```diff
@@ -40,10 +40,10 @@ export const App = ({ Navbar, Footer }) => {
       <Container>
         {Navbar && <Navbar user={user} isBESAdmin={isBESAdmin} />}
         <Switch>
-          <Route path={`${basePath}/viz-builder/:vizType/new`} exact>
+          <Route path={`${basePath}/viz-builder/:dashboardItemOrMapOverlay/new`} exact>
```
Comment (biaoli0) [inline packages/admin-panel/src/VizBuilderApp/App.js:43]: I can see why `vizType` is confusing. As there are dashboard item types too.

File: packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js [hunk 2]
```diff
@@ -141,11 +142,13 @@ export const PreviewSection = () => {
   const { hasPresentationError, setPresentationError } = useVizConfigError();
 
   const [{ project, location, testData }, { setPresentation }] = useVizConfig();
-  const { visualisationForFetchingData: visualisation } = useVisualisation();
+  const [{ visualisation }, { setVisualisationValue }] = useVizConfig();
```
Comment (biaoli0) [inline packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js:145]:
```suggestion
  const [{ project, location, testData, visualisation }, { setPresentation, setVisualisationValue }] = useVizConfig();
```

File: packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js [hunk 3]
```diff
@@ -230,11 +263,16 @@ export const PreviewSection = () => {
               <IdleMessage />
             )}
           </ChartContainer>
+          <VizTypeSelector value={vizType} onChange={newVizType => setVizType(newVizType)} />
```
Comment (biaoli0) [inline packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js:266]:
```suggestion
          <VizTypeSelector value={vizType} onChange={setVizType} />
```

**Review by biaoli0 — APPROVED** [review]
@IgorNadj Thanks for adding this validation! Stepping on this, and one more step we can get rid of json editor for presentation options 🥇 .

File: packages/ui-components/package.json [hunk 4]
```diff
@@ -27,10 +27,11 @@
     "@material-ui/icons": "^4.9.1",
     "@material-ui/lab": "^4.0.0-alpha.47",
     "@material-ui/pickers": "^3.2.10",
+    "@trevoreyre/autocomplete-js": "^2.4.1",
```
[the remark this answers is not in the document]
Comment (biaoli0) [reply packages/ui-components/package.json:30]: Probably this is required or a dependency of a library?

File: packages/types/src/types/models-extra/dashboard-item/common.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+   * @description A short description that appears above a viz
+   */
+  description?: string;
+
+  /**
+   * @description A comment on the viz that doesn't show up anywhere
+   */
+  '//'?: string;
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/types/src/types/models-extra/dashboard-item/common.ts:17]: Yeah I found it in a couple configs... may be better to remove it from the type definition so people don't use it.

File: packages/ui-components/src/components/Inputs/GroupedSelect.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            {item.label}
+          </MenuItem>
+        ),
+      )}
+    </GroupedSelectField>
+  );
+};
+
+GroupedSelect.propTypes = {
+  id: PropTypes.string.isRequired,
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/ui-components/src/components/Inputs/GroupedSelect.js:114]: Not sure, I just copy-pasted the Select component.

File: packages/ui-components/src/components/Inputs/GroupedSelect.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+export const MenuItem = styled(MuiMenuItem)`
+  padding-top: 0.75rem;
+  padding-bottom: 0.5rem;
+`;
+
+export const GroupedSelect = ({
+  value,
+  onChange,
+  groupedOptions,
+  showPlaceholder,
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/ui-components/src/components/Inputs/GroupedSelect.js:60]: Same as above, copy pasted from Select.js.

File: packages/ui-components/src/components/Inputs/GroupedSelect.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    >
+      {showPlaceholder && (
+        <MenuItem value="" disabled>
+          {placeholder}
+        </MenuItem>
+      )}
+
+      {flatSetOfItems.map(item =>
+        item.type === 'heading' ? (
+          <ListSubheader>{item.label}</ListSubheader>
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/ui-components/src/components/Inputs/GroupedSelect.js:102]: 👍 Fixed

File: packages/ui-components/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/ui-components/package.json:30]: 👍 Removed

File: packages/types/src/types/models-extra/dashboard-item/common.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/types/src/types/models-extra/dashboard-item/common.ts:17]: Removed

File: packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js (hunk 2 above)
Comment (IgorNadj) [reply packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js:145]: 👍 Fixed

File: packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js (hunk 3 above)
Comment (IgorNadj) [reply packages/admin-panel/src/VizBuilderApp/components/PreviewSection.js:266]: Updated

**Review by biaoli0 — APPROVED** [review]

**Pull request merged**